### PR TITLE
add analyzer support for set Local

### PR DESCRIPTION
**Pull request by m-kharbat**
@chaudum please review

### CODE REVIEW

File: sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java [hunk 1]
```diff
@@ -56,10 +56,14 @@
 
     @Test
     public void testSet() throws Exception {
-        SetAnalyzedStatement analysis = analyze("SET GLOBAL PERSISTENT stats.operations_log_size=1");
```
Comment (chaudum) [inline sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java]: why did you change the scope in the test?

File: sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java [hunk 2]
```diff
@@ -40,23 +40,42 @@
     private SetStatementAnalyzer() {}
 
     public static SetAnalyzedStatement analyze(SetStatement node, ParameterContext parameterContext) {
-        if (SetStatement.Scope.SESSION.equals(node.scope())) {
-            logger.debug("SET SESSION STATEMENT: {}", node.toString());
+
+        if (SetStatement.Scope.GLOBAL.equals(node.scope())) {
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java]: not necessary if handled in parser

File: sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java [hunk 3]
```diff
@@ -113,6 +133,39 @@ public void testSetInvalidStringValue() throws Exception {
     }
 
     @Test
+    public void testSetSessionInvalidArrayValue() throws Exception {
```
Comment (chaudum) [inline sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java]: the following 4 tests are not necessary if DEFAULT is handled in the parser

File: sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java [hunk 4]
```diff
@@ -40,23 +40,42 @@
...
+        } else {
+            for (Assignment assignment : node.assignments()){
+                if (assignment.expressions().contains(new QualifiedNameReference(new QualifiedName("default"))) && assignment.expressions().size() > 1){
+                    throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                        "DEFAULT cannot be used in a list assignment"));
+                }
+            }
+
+
+            logger.debug("SET {} STATEMENT: {}",node.scope().toString() , node.toString());
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java]: should be loglevel `WARN`

Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java]: there is also already a `SetSessionPlan` (which results in a `NoopTask` tho)
guess `TransportExecutor.visitSetPlan()` is the better place to log the warning

File: sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java [hunk 5]
```diff
@@ -71,6 +75,22 @@ public void testSet() throws Exception {
...
+
+        analysis = analyze("SET LOCAL stats.jobs_log_size TO 2");
+        assertThat(analysis.settings().toDelimitedString(','), is("stats.jobs_log_size=[2],"));
+
+        analysis = analyze("SET stats.jobs_log_size TO 2");
+        assertThat(analysis.settings().toDelimitedString(','), is("stats.jobs_log_size=[2],"));
+
+        analysis = analyze("SET SESSION stats.jobs_log_size = 1,2,3");
+        assertThat(analysis.settings().toDelimitedString(','), is("stats.jobs_log_size=[1, 2, 3],"));
+
```
Comment (kovrus) [inline sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java]: empty line

File: sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java [hunk 6]
```diff
@@ -40,23 +40,23 @@
     private SetStatementAnalyzer() {}
 
     public static SetAnalyzedStatement analyze(SetStatement node, ParameterContext parameterContext) {
-        if (SetStatement.Scope.SESSION.equals(node.scope())) {
-            logger.debug("SET SESSION STATEMENT: {}", node.toString());
+        if (!SetStatement.Scope.GLOBAL.equals(node.scope())) {
+            logger.warn("SET {} STATEMENT: {}", node.scope().toString(), node.toString());
```
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java]: The node.toString() already prints out the scope, so printing it again is not necessary

File: sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java [hunk 7]
```diff
@@ -60,6 +60,10 @@ public void testSet() throws Exception {
         assertThat(analysis.isPersistent(), is(true));
         assertThat(analysis.settings().toDelimitedString(','), is("stats.operations_log_size=1,"));
 
+        analysis = analyze("SET GLOBAL TRANSIENT stats.jobs_log_size='1'");
```
Comment (matriv) [inline sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java]: You can use the test below with multiple assignments to test '1' and 1 and remove this.

File: sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java [hunk 8]
```diff
@@ -71,6 +75,21 @@ public void testSet() throws Exception {
                 .put("stats.operations_log_size", 0)
                 .put("stats.jobs_log_size", 0)
                 .build()));
+
+        analysis = analyze("SET LOCAL stats.jobs_log_size TO 2");
```
Comment (matriv) [inline sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java]: add assertions for the scope here and below

File: sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java [hunk 9]
```diff
@@ -71,6 +75,21 @@ public void testSet() throws Exception {
...
+        assertThat(analysis.settings().toDelimitedString(','), is("stats.jobs_log_size=[2],"));
+
+        analysis = analyze("SET stats.jobs_log_size TO 2");
+        assertThat(analysis.settings().toDelimitedString(','), is("stats.jobs_log_size=[2],"));
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/SetAnalyzerTest.java]: maybe split these tests into multiple test functions.
Makes it more clear what aspect is being tested and if something breaks it will also be easier to debug

File: sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java [hunk 10]
```diff
@@ -40,23 +40,23 @@
     private SetStatementAnalyzer() {}
 
     public static SetAnalyzedStatement analyze(SetStatement node, ParameterContext parameterContext) {
-        if (SetStatement.Scope.SESSION.equals(node.scope())) {
-            logger.debug("SET SESSION STATEMENT: {}", node.toString());
+        if (!SetStatement.Scope.GLOBAL.equals(node.scope())) {
+            logger.warn("SET STATEMENT: {}", node);
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java]: the message does not say anything ...

File: sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java [hunk 11]
```diff
@@ -40,23 +40,23 @@
...
-            } else {
+            if (SetStatement.Scope.GLOBAL.equals(node.scope())) {
                 SettingsApplier settingsApplier = CrateSettings.getSettingsApplier(settingsName);
                 for (String setting : ExpressionToSettingNameListVisitor.convert(assignment)) {
                     checkIfSettingIsRuntime(setting);
                 }
-
                 settingsApplier.apply(builder, parameterContext.parameters(), assignment.expression());
+            } else {
+                builder.put(settingsName, assignment.expressions());
```
Comment (chaudum) [inline sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java]: we also need to throw an exception if the settingName exists in GLOBAL scope, but is used in LOCAL or SESSIOn scope.

File: sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java [hunk 12]
```diff
@@ -40,23 +40,23 @@
...
-                builder.put(settingsName, assignment.expressions());
-            } else {
+            if (SetStatement.Scope.GLOBAL.equals(node.scope())) {
                 SettingsApplier settingsApplier = CrateSettings.getSettingsApplier(settingsName);
                 for (String setting : ExpressionToSettingNameListVisitor.convert(assignment)) {
                     checkIfSettingIsRuntime(setting);
                 }
-
                 settingsApplier.apply(builder, parameterContext.parameters(), assignment.expression());
+            } else {
```
Comment (m-kharbat) [inline sql/src/main/java/io/crate/analyze/SetStatementAnalyzer.java]: it is used in the case of `SESSION` and `LOCAL`
